### ISSUE TITLE
Adding CURRENT_TIMESTAMP as a "DEFAULT" and "ON UPDATE" for createdAt, updatedAt on migrations

### ISSUE THREAD

**Issue by dorment** [sequelize/sequelize#4896]
I am using migrations to create and update my MySQL DB schemas. I am using underscored versions of createdAt, updatedAt. Here is what I have in my migration file:

      created_at: {
        allowNull: false,
        type: Sequelize.DATE
      },
      updated_at: {
        allowNull: false,
        type: Sequelize.DATE
      }

When I run the migration, I would expect "DEFAULT" to be set to CURRENT_TIMESTAMP for "created_at" and "ON UPDATE" to be set to CURRENT_TIMESTAMP for "updated_at", but it doesn't happen.

So, I am trying to set them myself. I have tried adding:

defaultValue: Sequelize.NOW
updatedAt: Sequelize.NOW
defaultValue: Sequelize.fn('NOW')
updatedAt: Sequelize.fn('NOW')
defaultValue: 'CURRENT_TIMESTAMP'
updatedAt: 'CURRENT_TIMESTAMP'
default: Sequelize.NOW

...all to no avail. I can't find any way to get "DEFAULT CURRENT_TIMESTAMP" or "ON UPDATE CURRENT_TIMESTAMP" to run in the query.

I know that sequelize handles the create and update times automatically, but it would be nice to be able to also put these in the raw SQL to allow SQL to handle them (in the case of manually making edits to the database, etc.)

Thanks!

**Comment by mickhansen** [comment]
`defaultValue: sequelize.literal('CURRENT_TIMESTAMP)` might work, not sure though.

**Comment by dorment** [comment]
That worked beautifully! Thanks!

Still trying to figure out the "ON UPDATE CURRENT_TIMESTAMP" case, if you have any ideas.

**Comment by mickhansen** [comment]
Does that work with default value or is it an extra field attribute?

**Comment by dorment** [comment]
It is an "Extra" field attribute.
![screenshot](https://cloud.githubusercontent.com/assets/7131945/11348332/2b3bffe2-91e3-11e5-8944-218e8754f70f.png)

**Comment by mickhansen** [comment]
Hmmm, then that is likely not possible with native sequelize at the moment.

**Comment by dorment** [comment]
OK, thanks!

**Comment by shun-tak** [comment]
I tried following:

```
queryInterface.createTable(
    'user',
    {
        'id': {
            type: Sequelize.BIGINT,
            primaryKey: true,
            autoIncrement: true,
        },
        'name': {
            type: Sequelize.STRING,
        },
        'created_at': {
            type: Sequelize.DATE(3),
            defaultValue: Sequelize.literal('CURRENT_TIMESTAMP(3)'),
        },
        'updated_at': {
            type: Sequelize.DATE(3),
            defaultValue: Sequelize.literal('CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3)'),
        },
    },
    {
        engine: 'InnoDB',
        charset: 'utf8mb4',
    }
);
```

and I got this table:

```
CREATE TABLE `user` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `cognito_identity_id` varchar(255) DEFAULT NULL,
  `created_at` datetime(3) DEFAULT CURRENT_TIMESTAMP(3),
  `updated_at` datetime(3) DEFAULT CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3),
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```

My env is `Sequelize [Node: 4.3.2, CLI: 2.4.0, ORM: 3.27.0, mysql: ^2.12.0]`

Although it is closed :)

**Comment by Here21** [comment]
@shun-tak Thanks man!
```js

module.exports = function (sequelize, DataTypes) {
  const Model = sequelize.define('TUser', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
      unique: true,
      field: 'id',
    },
    'created_at': {
      type: DataTypes.DATE(3),
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP(3)'),
    },
    'updated_at': {
      type: DataTypes.DATE(3),
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3)'),
    },
  }, {
    timestamps: true,
    tableName: 't_user',
    paranoid: true,
    underscored: true,
  });
  return Model
};
```

**Comment by kartik115** [comment]
But how it will act while insertion. Sequelize orm showing error 
INSERT INTO `user` (`id`,`name`,`created_at`,`updated_at`,`user_role_id`,`company_id`) VALUES (DEFAULT,'Kartikeya Mishra',CURRENT_TIMESTAMP,CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,1,1);

**Comment by kartik115** [comment]
SequelizeDatabaseError: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'ON UPDATE CURRENT_TIMESTAMP,1,1)

**Comment by shun-tak** [comment]
@kartik115 Just insert like this:
```
INSERT INTO user (id,name) VALUES (DEFAULT,'Alice');
```
or
```
INSERT INTO user (id,name,created_at,updated_at) VALUES (DEFAULT,'Alice',CURRENT_TIMESTAMP,CURRENT_TIMESTAMP);
```

Please see the MySQL manual.
https://dev.mysql.com/doc/refman/5.6/en/timestamp-initialization.html

**Comment by kartik115** [comment]
@shun-tak I got this error when I was using sequelize orm to create a new user. So it showing me syntax error. I have used 
'updated_at': {
            type: Sequelize.DATE(3),
            defaultValue: Sequelize.literal('CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3)'),
        },
So while creating new user its using the default value "ON UPDATE CURRENT_TIMESTAMP" and showing me error.

**Comment by lvnar** [comment]
Same error. Message:
```
String based operators are now deprecated. Please use Symbol based operators for better security.
SequelizeDatabaseError: syntax error at or near "ON".
```
I've read about Sequelize operators on querying, but there's nothing abou 'ON'. :(

**Comment by mrfoe7** [comment]
@kartik115 using in sequelize version 3.21.0,
'updated_at': {
type: "TIMESTAMP"
}.
I don't know how, but it is working 🥁

**Comment by amwill04** [comment]
Anyone coming here in 2018 @shun-tak method no longer works for `sequelize v4` when you try to use `update` method.

`queryInterface.createTable(
    'user',
    {
        'id': {
            type: Sequelize.BIGINT,
            primaryKey: true,
            autoIncrement: true,
        },
        'name': {
            type: Sequelize.STRING,
        },
        'created_at': {
            type: Sequelize.DATE(3),
            defaultValue: Sequelize.literal('CURRENT_TIMESTAMP(3)'),
        },
        'updated_at': {
            type: Sequelize.DATE(3),
            defaultValue: Sequelize.literal('CURRENT_TIMESTAMP(3)'),
        },
    },
    {
        engine: 'InnoDB',
        charset: 'utf8mb4',
    }
);`

Basically remove the 'ON UPDATE CURRENT_TIMESTAMP(3)' otherwise this will result in SQL syntax error when you try to update. Like such:

`UPDATE Users SET firstName='Bob', updatedAt=CURRENT_TIMESTAMP
(3) ON UPDATE CURRENT_TIMESTAMP(3) WHERE id = 1`

**Comment by tomascharad** [comment]
+1

**Comment by tomascharad** [comment]
@amwill04, Does your solution updates the updatedAt attribute automatically?

**Comment by brentwashburne** [comment]
In version 4.38.0, the file sequelize/lib/dialects/mysql/query-generator.js uses `attribute.onUpdate` at line 372, but that is inside a block at line 351: `if (attribute.references) {`.  In other words, `onUpdate` is only used when there is a `references` attribute too.  This is a bug, `onUpdate` should be allowed without a reference to another model.

**Comment by cloudge** [comment]
It works for me
```JavaScript
queryInterface.createTable('User', {
      id: {
        allowNull: false,
        autoIncrement: true,
        primaryKey: true,
        type: Sequelize.INTEGER
      },
      name: {
        type: Sequelize.STRING
      },
      createdAt: {
        allowNull: false,
        type: Sequelize.DATE,
        defaultValue: Sequelize.literal('CURRENT_TIMESTAMP')
      },
      updatedAt: {
        allowNull: false,
        type: Sequelize.DATE,
        defaultValue: Sequelize.literal('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP'),
      }
    });
```
![image](https://user-images.githubusercontent.com/8080412/49245120-51db2700-f409-11e8-8fe4-118d9c5906be.png)

**Comment by mrfoe7** [comment]
@cloudge please, tell about you sequelize version.

**Comment by aqellez** [comment]
All variants doesn't work correctly for me in postgres. They are INVOKED (WHY?!) and timestamp is setted to time of model's script execution. Not literal, not simple string 'now', nothing else doesn't work.

**Comment by shakti-paktolus** [comment]
> `defaultValue: sequelize.literal('CURRENT_TIMESTAMP)` might work, not sure though.

Only solution which actually works when updated with both code and DB directly. Kudos.

**Comment by krushndayshmookh** [comment]
> sequelize.literal('CURRENT_TIMESTAMP)

this solved my problem. thank you so much!

here's what i did earlier: 
```js
created_on: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: "CURRENT_TIMESTAMP(6)"
}
```
it was solved with:
```js
created_on: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: sequelize.literal("CURRENT_TIMESTAMP(6)")
}
```

**Comment by airglow923** [comment]
For those who are still confused, consider the code below:

First, use `Sequelize.literal('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP')` as mentioned above:

```js
const sample = (sequelize, Sequelize) =>
  sequelize.define("sample", {
    createdAt: {
      type: Sequelize.DATE,
      defaultValue: Sequelize.literal("CURRENT_TIMESTAMP"),
    },
    updatedAt: {
      type: Sequelize.DATE,
      defaultValue: Sequelize.literal(
        "CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP"
      ),
    },
  });
```

Then, use `COLUMN_NAME: Sequelize.literal("CURRENT_TIMESTAMP") ` to insert:

```js
const Sample = sample(sequelize, Sequelize);

Sample.create({ updatedAt: Sequelize.literal("CURRENT_TIMESTAMP") }).then(
  (instance) => {
    instance.save();
  }
);
```

**Comment by sajjadjaved01** [comment]
works for now.
```
createdAt: {
        allowNull: false,
        type: 'TIMESTAMP',
      },
```

**Comment by dishantdave21** [comment]
> @kartik115 using in sequelize version 3.21.0, 'updated_at': { type: "TIMESTAMP" }. I don't know how, but it is working 🥁

ERROR: TIMESTAMP is not defined

"dependencies": {
     "mysql2": "^2.3.3",
    "sequelize": "^6.21.0",
    "sequelize-cli": "^6.4.1"
}

**Comment by SiranWeb** [comment]
Just a quick note who also spent a lot of time as me to make `updatedAt` field work. `ON UPDATE` trigger works in **MySQL**, but not in **PostgreSQL**. Sequelize updates `updatedAt` field by itself. So, it's would be enough if you just write these lines in your migration:
```js
updatedAt: {
      type: Sequelize.DATE,
      defaultValue: Sequelize.literal("CURRENT_TIMESTAMP"),
    }
```

**Comment by EmileIB** [comment]
> It works for me
> 
> ```js
> queryInterface.createTable('User', {
>       id: {
>         allowNull: false,
>         autoIncrement: true,
>         primaryKey: true,
>         type: Sequelize.INTEGER
>       },
>       name: {
>         type: Sequelize.STRING
>       },
>       createdAt: {
>         allowNull: false,
>         type: Sequelize.DATE,
>         defaultValue: Sequelize.literal('CURRENT_TIMESTAMP')
>       },
>       updatedAt: {
>         allowNull: false,
>         type: Sequelize.DATE,
>         defaultValue: Sequelize.literal('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP'),
>       }
>     });
> ```
> 
> ![image](https://user-images.githubusercontent.com/8080412/49245120-51db2700-f409-11e8-8fe4-118d9c5906be.png)

what version are you using?

**Comment by faheel** [comment]
The suggestion by @miladabc in #5561 was the only one that worked both in the app and the DB (i.e. no matter where the entry was inserted/updated, the right timestamps were always set).

See https://github.com/sequelize/sequelize/issues/5561#issuecomment-1772625328 for more details

**Comment by dersimn** [comment]
> It works for me
> 
> ```js
> queryInterface.createTable('User', {
>       id: {
>         allowNull: false,
>         autoIncrement: true,
>         primaryKey: true,
>         type: Sequelize.INTEGER
>       },
>       name: {
>         type: Sequelize.STRING
>       },
>       createdAt: {
>         allowNull: false,
>         type: Sequelize.DATE,
>         defaultValue: Sequelize.literal('CURRENT_TIMESTAMP')
>       },
>       updatedAt: {
>         allowNull: false,
>         type: Sequelize.DATE,
>         defaultValue: Sequelize.literal('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP'),
>       }
>     });
> ```
> 
> ![image](https://user-images.githubusercontent.com/8080412/49245120-51db2700-f409-11e8-8fe4-118d9c5906be.png)

Works as long as you don't use automated insertion:

```
const User = sequelize.define('user', {
    name: DataTypes.TEXT,
    favoriteColor: {
        type: DataTypes.TEXT,
        defaultValue: 'green',
    },
    age: DataTypes.INTEGER,
    cash: DataTypes.INTEGER,
    createdAt: {
        type: Sequelize.DATE,
        defaultValue: Sequelize.literal('CURRENT_TIMESTAMP')
    },
    updatedAt: {
        type: Sequelize.DATE,
        defaultValue: Sequelize.literal('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP')
    }
})
await sequelize.sync()

// works till here

const jane = await User.create({ name: 'Jane' })  // fails here
```

**Comment by ankit12332** [comment]
> defaultValue: Sequelize.literal('CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3)'),

in here
When I am inserting the data. it's updating currentdate and updateddate as well
Why it's updating date while inserting
